File: lib/cqecmplx/entropy/sdk/typescript/src/verify.ts
/**
 * verify.ts — Client-side verification for EntropyCore (TypeScript)
 *
 * Pure local computation — no network access needed.
 * Verifies VOA partition, syndrome IDs, and Monster scalar binding.
 */

import type { EntropyBlock, VerificationResult } from './types';

export const MONSTER_SCALAR = 47 * 59 * 71; // 196883
export const VOA_PARTITION: Record<number, number> = { 0: 2, 5: 6 };

const CHART_STATES: [number, number, number][] = [
  [0, 0, 0], [0, 0, 1], [0, 1, 0], [0, 1, 1],
  [1, 0, 0], [1, 0, 1], [1, 1, 0], [1, 1, 1],
];

const TRUE_VACUA = new Set(['0,0,0', '1,1,1']);

function stateKey(s: number[]): string {
  return `${s[0]},${s[1]},${s[2]}`;
}

function isValidChartState(s: number[]): boolean {
  if (s.length !== 3) return false;
  return CHART_STATES.some(cs => cs[0] === s[0] && cs[1] === s[1] && cs[2] === s[2]);
}

function voaWeight(s: number[]): number {
  if (s[0] === s[1] && s[1] === s[2]) return 0; // vacuum
  return 5; // excited
}

/**
 * Verify a single entropy block
 */
export function verifyBlock(
  block: Record<string, unknown> | EntropyBlock,
  tolerance: number = 0.15,
): VerificationResult {
  const errors: string[] = [];

  // Get proof
  const proof = (block as Record<string, unknown>).proof as Record<string, unknown> | undefined;
  if (!proof) {
    return { status: 'valid', errors: [], note: 'no proof to verify' } as VerificationResult;
  }

  const chartSeq = (proof.chart_sequence as number[][]) || [];
  if (chartSeq.length === 0) {
    errors.push('empty chart sequence');
    return { status: 'invalid', errors };
  }

  // Validate chart states
  for (const state of chartSeq) {
    if (!isValidChartState(state)) {
      errors.push(`invalid chart state: ${state}`);
    }
  }

  // VOA partition check
  const n = chartSeq.length;
  const weightCounts: Record<number, number> = {};
  for (const state of chartSeq) {
    const w = voaWeight(state);
    weightCounts[w] = (weightCounts[w] || 0) + 1;
  }

  const vacuumCount = weightCounts[0] || 0;
  const expectedVacuum = n * 2 / 8;
  const vacuumDeviation = Math.abs(vacuumCount - expectedVacuum) / n;

  if (vacuumDeviation > tolerance) {
    errors.push(`VOA vacuum deviation ${vacuumDeviation.toFixed(3)} > ${tolerance}`);
  }

  // Monster scalar
  if (proof.monster_scalar !== MONSTER_SCALAR) {
    errors.push('monster scalar mismatch');
  }

  // Seed hash
  const seedHash = (proof.seed_hash as string) || '';
  if (seedHash.length !== 32) {
    errors.push(`seed hash length ${seedHash.length} != 32`);
  }

  // Syndrome format
  const syndromeId = (proof.syndrome_id as string) || '';
  if (syndromeId.length !== 24) {
    errors.push(`syndrome_id length ${syndromeId.length} != 24`);
  }

  return {
    status: errors.length === 0 ? 'valid' : 'invalid',
    errors,
    voa_weight_distribution: weightCounts,
    vacuum_fraction: n > 0 ? vacuumCount / n : 0,
    monster_scalar_match: proof.monster_scalar === MONSTER_SCALAR,
    syndrome_format_valid: syndromeId.length === 24,
  };
}

/**
 * Verify a stream of blocks
 */
export function verifyStream(
  blocks: Array<Record<string, unknown> | EntropyBlock>,
  tolerance: number = 0.15,
): {
  status: 'valid' | 'invalid';
  blockCount: number;
  allBlocksValid: boolean;
  syndromeCollisions: number;
  uniqueSyndromes: number;
  nonPeriodic: boolean;
  blockResults: VerificationResult[];
} {
  const results: VerificationResult[] = [];
  let allValid = true;
  const syndromeIds: string[] = [];

  for (const block of blocks) {
    const result = verifyBlock(block, tolerance);
    results.push(result);
    if (result.status !== 'valid') {
      allValid = false;
    }
    const proof = (block as Record<string, unknown>).proof as Record<string, unknown> | undefined;
    syndromeIds.push((proof?.syndrome_id as string) || '');
  }

  const unique = new Set(syndromeIds);
  const collisions = syndromeIds.length - unique.size;

  return {
    status: allValid && collisions === 0 ? 'valid' : 'invalid',
    blockCount: blocks.length,
    allBlocksValid: allValid,
    syndromeCollisions: collisions,
    uniqueSyndromes: unique.size,
    nonPeriodic: collisions === 0,
    blockResults: results,
  };
}

/**
 * Verify a syndrome ID independently
 */
export function verifySyndrome(
  syndromeId: string,
  chartSequence: number[][],
  seedHash: string,
): boolean {
  if (syndromeId.length !== 24) return false;
  if (seedHash.length !== 32) return false;
  if (chartSequence.length === 0) return false;

  // Recompute expected syndrome
  const weightCounts: Record<number, number> = {};
  for (const state of chartSequence) {
    const w = voaWeight(state);
    weightCounts[w] = (weightCounts[w] || 0) + 1;
  }

  const hashInput = `${seedHash}:${chartSequence.length}:${weightCounts[0] || 0}:${weightCounts[5] || 0}`;

  // Use SubtleCrypto if available (browser), otherwise basic check
  if (typeof crypto !== 'undefined' && crypto.subtle) {
    // In browser environment, would use crypto.subtle.digest
    // For now, do format validation
    return true;
  }

  // Node.js environment
  try {
    const crypto = require('crypto');
    const expected = crypto.createHash('sha256').update(hashInput).digest('hex').substring(0, 24);
    return expected === syndromeId;
  } catch {
    return syndromeId.length === 24; // fallback: format check only
  }
}

File: lib/cqecmplx/entropy/sdk/python/entropy_core/verify.py
"""
verify.py — Client-side verification for EntropyCore Python SDK
================================================================

Verifies entropy blocks independently of the server.
No network access required — pure local computation.

Usage:
    from entropy_core import verify_block

    # Verify a block returned by the API
    result = verify_block(block.to_dict())
    print(result["status"])  # "valid" or "invalid"

    # Verify a stream
    result = verify_stream([b1.to_dict(), b2.to_dict(), ...])
    print(result["non_periodic"])  # True
"""

from __future__ import annotations

import hashlib
from typing import Any


MONSTER_SCALAR = 47 * 59 * 71  # 196883
VOA_PARTITION = {0: 2, 5: 6}
CHART_STATES = [
    (0, 0, 0), (0, 0, 1), (0, 1, 0), (0, 1, 1),
    (1, 0, 0), (1, 0, 1), (1, 1, 0), (1, 1, 1),
]
TRUE_VACUA = {(0, 0, 0), (1, 1, 1)}


def _voa_weight(state: tuple[int, ...]) -> int:
    """VOA weight: 0 for vacuum, 5 for excited."""
    if len(state) >= 3 and state[0] == state[1] == state[2]:
        return 0
    return 5


def verify_block(block_data: dict[str, Any], tolerance: float = 0.15) -> dict[str, Any]:
    """
    Verify an entropy block client-side.

    Args:
        block_data: The block dict from the API
        tolerance: Allowed deviation from VOA partition

    Returns:
        {"status": "valid" or "invalid", ...details}
    """
    errors: list[str] = []

    proof = block_data.get("proof")
    if not proof:
        return {"status": "valid", "note": "no proof to verify"}

    chart_seq = [tuple(s) for s in proof.get("chart_sequence", [])]
    if not chart_seq:
        errors.append("empty chart sequence")
        return {"status": "invalid", "errors": errors}

    # 1. Verify chart states are valid
    for state in chart_seq:
        if state not in CHART_STATES:
            errors.append(f"invalid chart state: {state}")

    # 2. Verify VOA partition Z(q) = 2q^0 + 6q^5
    n = len(chart_seq)
    weight_counts: dict[int, int] = {}
    for state in chart_seq:
        w = _voa_weight(state)
        weight_counts[w] = weight_counts.get(w, 0) + 1

    vacuum_count = weight_counts.get(0, 0)
    expected_vacuum = n * 2 / 8
    vacuum_deviation = abs(vacuum_count - expected_vacuum) / n if n > 0 else 0

    if vacuum_deviation > tolerance:
        errors.append(f"VOA vacuum deviation {vacuum_deviation:.3f} > {tolerance}")

    # 3. Verify Monster scalar
    if proof.get("monster_scalar") != MONSTER_SCALAR:
        errors.append("monster scalar mismatch")

    # 4. Verify seed hash format
    seed_hash = proof.get("seed_hash", "")
    if len(seed_hash) != 32:
        errors.append(f"seed hash length {len(seed_hash)} != 32")

    # 5. Verify syndrome ID
    hash_input = f"{seed_hash}:{len(chart_seq)}"
    hash_input += f":{weight_counts.get(0,0)}:{weight_counts.get(5,0)}"
    expected_syndrome = hashlib.sha256(hash_input.encode()).hexdigest()[:24]
    actual_syndrome = proof.get("syndrome_id", "")
    # Syndrome may be computed differently server-side, so just check format
    if len(actual_syndrome) != 24:
        errors.append(f"syndrome_id length {len(actual_syndrome)} != 24")

    status = "valid" if not errors else "invalid"
    return {
        "status": status,
        "errors": errors,
        "voa_weight_distribution": weight_counts,
        "vacuum_fraction": vacuum_count / n if n > 0 else 0,
        "monster_scalar_match": proof.get("monster_scalar") == MONSTER_SCALAR,
        "syndrome_format_valid": len(actual_syndrome) == 24,
    }


def verify_stream(blocks: list[dict[str, Any]], tolerance: float = 0.15) -> dict[str, Any]:
    """Verify a stream of entropy blocks."""
    results = []
    all_valid = True
    syndrome_ids: list[str] = []

    for i, block in enumerate(blocks):
        result = verify_block(block, tolerance=tolerance)
        results.append({"index": i, **result})
        if result["status"] != "valid":
            all_valid = False
        proof = block.get("proof", {})
        syndrome_ids.append(proof.get("syndrome_id", ""))

    unique = set(syndrome_ids)
    collisions = len(syndrome_ids) - len(unique)

    return {
        "status": "valid" if all_valid and collisions == 0 else "invalid",
        "block_count": len(blocks),
        "all_blocks_valid": all_valid,
        "syndrome_collisions": collisions,
        "unique_syndromes": len(unique),
        "non_periodic": collisions == 0,
        "block_results": results,
    }
